### DISCUSSION TITLE
Converting Python Speech Recognition Audio Frame data to a numpy array that can be processed by Whisper?

### DISCUSSION THREAD

**Discussion by CareNetAI** [openai/whisper#2051] [Q&A] [answered]
I'm using the speech recognition Python library to record audio bytes from my microphone in mono at 16khz but I want to use the new Whisper library that accepts NumPy arrays, spectrograms, and file paths. Writing to a file takes too long so I'd like to directly convert the data to an array to pass it to Whisper.

**Comment by Mattral** [comment, marked as answer]
Here is a solution to your problem:

Assuming your code goes like:
```
with sr.Microphone(device_index=device_index, sample_rate=16000) as source:
    r = sr.Recognizer()
    audio = r.listen(source, timeout=None)
```

you need to convert the audio data (the output of your Recognizer.listen) to wave format like in this reference [1](https://stackoverflow.com/questions/46738569/wav-file-from-microphone-on-python)

```
audio_data = audio.get_wav_data()
```

which can be converted to an array of int16 [2](https://stackoverflow.com/questions/66296066/python-3-convert-wave-data-byte-array-to-numpy-array-of-floating-point-values)

```
data_s16 = np.frombuffer(audio_data, dtype=np.int16, count=len(audio_data)//2, offset=0)
```

which can then be converted to an array of float32 

```
float_data = data_s16.astype(np.float32, order='C') / 32768.0
```

which can then be processed by whisper. If there is a faster way let me know.

**Reply by CareNetAI** [reply]
Thanks for the advice
